File: clojure/src/tetrisanalyzer/piece.clj
(ns tetrisanalyzer.piece)

;; -
(def A {:index 0 :char \- })

;; I
(def I0 [[0 0][1 0][2 0][3 0]])
(def I1 [[0 0][0 1][0 2][0 3]])
(def I {:index 1 :char \I :widths [3 2 3 2] :heights [2 3 2 3] :rotations [I0 I1]})

;; Z
(def Z0 [[0 0][1 0][1 1][2 1]])
(def Z1 [[1 0][0 1][1 1][0 2]])
(def Z {:index 2 :char \Z :widths [3 2] :heighs [2 3] :rotations [Z0 Z1]})

;; S
(def S0 [[1 0][2 0][0 1][1 1]])
(def S1 [[0 0][0 1][1 1][1 2]])
(def S {:index 3 :char \S :widths [3 2] :heiths [2 3] :rotations [S0 S1]})

;; J
(def J0 [[0 0][1 0][2 0][2 1]])
(def J1 [[0 0][1 0][0 1][0 2]])
(def J2 [[0 0][0 1][1 1][2 1]])
(def J3 [[1 0][1 1][0 2][1 2]])
(def J {:index 4, :char \J :widths [3 2 3 2] :heights [2 3 2 3] :rotations [J0 J1 J2 J3]})

;; L
(def L0 [[0 0][1 0][2 0][0 1]])
(def L1 [[0 0][0 1][0 2][1 2]])
(def L2 [[2 0][0 1][1 1][2 1]])
(def L3 [[0 0][1 0][1 1][1 2]])
(def L {:index 5, :char \L :widths [3 2 3 2] :heights [2 3 2 3] :rotations [L0 L1 L2 L3]})

;; T
(def T0 [[0 0][1 0][2 0][1 1]])
(def T1 [[0 0][0 1][1 1][0 2]])
(def T2 [[1 0][0 1][1 1][2 1]])
(def T3 [[1 0][0 1][1 1][1 2]])
(def T {:index 6, :char \T :widths [3 2 3 2] :heights [2 3 2 3] :rotations [T0 T1 T2 T3]})

;; O
(def O0 [[0 0][1 0][0 1][1 1]])
(def O {:index 7 :char \O :widths [2] :heiths [2] :rotations [O0]})

;; X
(def X {:index 8 :char \x })

(def pieces [A I Z S J L T O X])

(def char->piece {\- 0 \I 1 \Z 2 \S 3 \J 4 \L 5 \T 6 \O 7 \x 8})

(defn piece->char [piece] (nth "-IZSJLTOx" piece))

File: langs/java/src/main/java/com/github/tetrisanalyzer/piece/Piece.java
package com.github.tetrisanalyzer.piece;

import com.github.tetrisanalyzer.settings.PieceSettings;
import com.github.tetrisanalyzer.settings.adjustment.Adjustments;

import java.util.HashMap;
import java.util.Map;

/**
 * Represents one of the pieces: O, I, S, Z, L, J, T
 */
public abstract class Piece {
    private static final int[] ROTATION_MODULUS = { 0, 1, 0, 3 };

    public static final int NUMBER_OF_PIECE_TYPES = 7;

    public static final int O = 1;
    public static final int I = 2;
    public static final int S = 3;
    public static final int Z = 4;
    public static final int L = 5;
    public static final int J = 6;
    public static final int T = 7;

    public static PieceEmpty createPieceEmpty(PieceSettings settings) { return new PieceEmpty(settings); }
    public static PieceO createPieceO(PieceSettings settings) { return new PieceO(settings); }
    public static PieceI createPieceI(PieceSettings settings) { return new PieceI(settings); }
    public static PieceS createPieceS(PieceSettings settings) { return new PieceS(settings); }
    public static PieceZ createPieceZ(PieceSettings settings) { return new PieceZ(settings); }
    public static PieceL createPieceL(PieceSettings settings) { return new PieceL(settings); }
    public static PieceJ createPieceJ(PieceSettings settings) { return new PieceJ(settings); }
    public static PieceT createPieceT(PieceSettings settings) { return new PieceT(settings); }
    public static PieceAny createPieceAny(PieceSettings settings) { return new PieceAny(settings); }

    public static Piece[] validPieces(PieceSettings settings) {
        Piece[] pieces = new Piece[] {
                createPieceEmpty(settings),
                createPieceO(settings),
                createPieceI(settings),
                createPieceS(settings),
                createPieceZ(settings),
                createPieceL(settings),
                createPieceJ(settings),
                createPieceT(settings),
                createPieceAny(settings)
        };
        return pieces;
    }

    private static final Map<Character,Integer> indexMap = new HashMap<>();

    static {
        indexMap.put('-', 0);
        indexMap.put('O', 1);
        indexMap.put('I', 2);
        indexMap.put('S', 3);
        indexMap.put('Z', 4);
        indexMap.put('L', 5);
        indexMap.put('J', 6);
        indexMap.put('T', 7);
        indexMap.put('x', 8);
    }

    public static int indexOf(char piece) {
        return indexMap.get(piece);
    }

    public Piece(PieceSettings settings) {
        adjustments = settings.pieceAdjustments()[number()];
    }

    public final Adjustments adjustments;

    public abstract byte number();
    public abstract char character();
    public int rotationsEndIndex() { return heights().length - 1; }
    public int rotationModulus() { return ROTATION_MODULUS[rotationsEndIndex()]; }
    public int width(int rotation) { return widths()[rotation]; }
    public int height(int rotation) { return heights()[rotation]; }
    public int startX() { return adjustments.startX; }
    public int dx(int rotation) { return adjustments.dx.get(rotation); };
    public int dy(int rotation) { return adjustments.dy.get(rotation); };
    public boolean isAdjusted(int rotation) { return dx(rotation) != 0 || dy(rotation) != 0; }
    public PieceShape getShape(int rotation) { return shapes()[rotation]; }
    protected abstract int[] widths();
    protected abstract int[] heights();
    protected abstract PieceShape[] shapes();

    @Override
    public int hashCode() {
        return number();
    }

    @Override
    public boolean equals(Object obj) {
        if (!(obj instanceof Piece)) {
            return false;
        }
        return ((Piece)obj).number() == number();
    }

    @Override
    public String toString() {
        return String.valueOf(character());
    }
}
